### ISSUE TITLE
Could this theoretically be retrained from scratch to generate singing vocals?

### ISSUE THREAD

**Issue by Saltb0xApps** [huggingface/parler-tts#8]
Given a 10k hour dataset of singing vocals (_instead of the current audiobook reading content_), could this model be ported to be able to sing/generate vocals?

**Comment by adamfils** [comment]
@Saltb0xApps  I was thinking the same thing and also adding another input conditioning, like background music for the generated audio to follow. I am not exactly an ML engineer but here is my rough thinking.

1. Text Encoder (Unchanged):
Continues to map text descriptions to a sequence of hidden-state representations using a frozen text encoder initialized from Flan-T5.

2.Music Encoder:
A new component that takes background music as input and generates a music-conditioned representation using a pretrained music autoencoder (DAC or EnCodec).
This encoder analyses the background music to extract features such as tempo, key, mood, and rhythm, which will be used to condition the generated speech.

3. Parler-TTS Decoder (Modified):
The decoder now auto-regressively generates audio tokens conditional not only on the encoder hidden-state representations (from text) but also on the music-conditioned representation.
To incorporate the music-conditioned representation, you could either:
- Concatenate: Directly concatenate the music-conditioned representation with the text-conditioned hidden states before feeding them into the decoder.
- Cross-Attention Modification: Integrate the music-conditioned representation into the cross-attention layers of the decoder, allowing the decoder to attend to both text and music features simultaneously.

4. Audio Codec (Unchanged):
Continues to recover the audio waveform from the audio tokens predicted by the decoder using the DAC model or EnCodec as preferred.

**Comment by sanchit-gandhi** [comment]
Hey @Saltb0xApps @adamfils - this sounds like it would work. The only change that I would make would be using a more powerful audio encoder to extract more meaningful representations from the music conditioning (e.g. warm starting an audio encoder from the [HuBERT model](https://huggingface.co/docs/transformers/en/model_doc/hubert) to extract music embedding representations). Using DAC or EnCodec alone is only going to provide you with a down-sampled version of the music inputs, rather than any features that encode tempo, key, mood, and rhythm, etc. This is then analogous to what the Flan-T5 encoder does for the text conditioning.

Note that you could use something similar to train a TTS model that has text **and** voice conditioning as well (just replace the music conditioning with a voice sample in the flowchart above). You could then give it a 2-second voice prompt to control the style of generated voice, and then control how fast/slow or animated/monotonous the speech is using the text prompt.

**Comment by Saltb0xApps** [comment]
@sanchit-gandhi / @adamfils Thank you for providing a detailed response! 
I have a large dataset of around 1k hours of just vocals (_separated from music out of [Facebook's sound library](https://www.facebook.com/sound) using demucs_) and their lyrics/transcription using Whisper. I was wondering if I could take that dataset of just vocals, combined with dataspeech info for style, and retrain parler-tts to only output singing vocals.

The idea is to create a more robust singing version of [Suno AI's bark](https://github.com/suno-ai/bark) using parler-tts that can generate singing vocals instead of just regular speech.

Would this require code level changes as mentioned above, or simply retraining parler-tts on this dataset would be a good enough starting point for something that can generate vocals only?

**Comment by ylacombe** [comment]
Hey @Saltb0xApps, this would totally work as you only need 3 things from your datasets for the training to work:
1. Audio samples
2. Transcriptions
3. Text conditioning

Parler-TTS is agnostic to the text conditioning and audio samples you're using!

Also, 1K hours should be enough to get a good-enough model from scratch. You can also explore fine-tuning the current model as well as it already learned some acoustics features and to associate text tokens to acoustics sounds!

**Comment by Saltb0xApps** [comment]
@ylacombe I curated the dataset of 1000 hours of vocals (_mostly english_) 
Would love to hear your thoughts whether this dataset could work -
• audio + transcriptions - https://huggingface.co/datasets/AkhilTolani/vocals
• transcriptions + dataspeech tags - https://huggingface.co/datasets/AkhilTolani/vocals-stripped-pitch-text-bins-descriptions

I just started a finetune run here on 2xA100 GPUs. Would love to know if my fine-tune parameters are adjusted correctly for the difference in hardware/dataset size - https://wandb.ai/akhiltolani/parler-speech/runs/wkh5eor3/overview?nw=nwuserakhiltolani

**Comment by adamfils** [comment]
@Saltb0xApps awesome please share some of your audio outputs when you can as the training proceeds.

**Comment by ylacombe** [comment]
Hey @Saltb0xApps, wow thanks for sharing this! How did you create this dataset out of curiosity?

A few remarks:
- I'm pretty sure the model can learn from your samples but it would surely benefit from more precise tags: for now, it only uses the dataspeech tags but you would probably need more signals about the singing voices (for example, the singing style, the note etc). Maybe you can get some additional features from the way you created your dataset, would you like to describe this a bit more?
- You should also probably modify the [prompt](https://github.com/huggingface/dataspeech/blob/166ba7cad84fc145907ca1d0c77ad8ac3829cf9b/scripts/run_prompt_creation.py#L305-L315) to create descriptions a bit. The current one is about delivering speech, whereas you want to deliver singing.
- I'll give a proper look to the HPs ASAP

**Comment by ylacombe** [comment]
I've listened to some samples, the model seems to get a sense of singing, which is a good sign! It'd probably need some better hyper-parameters though!

Re: Hyper-parameters
- You've got 1k hours of audio, so you'd better raise your global batch size to get closer to the global batch size we've used for training (196), you can do it by setting gradient_accumulation_steps to something aound 6
- You might have to experiment with the learning rate, for now you can leave it as it is but it's a bit too high IMO
-

**Comment by Saltb0xApps** [comment]
> Thank you also for sharing your logs, they bring a lot of value to the community! I really like your initiative! If you're okay with this, we can probably make a big splash from your model once the results are what we expect. What do you think?

Absolutely! I'd love connect & discuss a plan over discord/email. Lmk if that works for you!

**Comment by Saltb0xApps** [comment]
> I've listened to some samples, the model seems to get a sense of singing, which is a good sign! It'd probably need some better hyper-parameters though!
> 
> Re: Hyper-parameters
> 
> * You've got 1k hours of audio, so you'd better raise your global batch size to get closer to the global batch size we've used for training (196), you can do it by setting gradient_accumulation_steps to something aound 6
> * You might have to experiment with the learning rate, for now you can leave it as it is but it's a bit too high IMO

@ylacombe  Ah thank you for sharing this! I definitely agree 1e-4 is too high. I'll do another run with 4e-5 as the lr and updated gradient_accumulation_steps to 6 later today.

**Comment by Saltb0xApps** [comment]
@ylacombe I believe parler-tts is going to come up with a larger model very soon if i remember correctly? 
600M is comparable to musicgen-small, and i believe there is a major qualitative difference between 3B+ params and 600M params (atleast in musicgen)! 
I'd imagine funetuning the parler-tts large model would most likely also give much better singing results!

**Comment by Saltb0xApps** [comment]
Here is a Huggingface space to try out the singing vocals fine-tune of Parler-tts! 
https://huggingface.co/spaces/AkhilTolani/vocals 

1. The model is having a very hard time differentiating between male and female vocals. Maybe its my training dataset?

2. Ideas on how to generate consistently long vocals based on speaker_id in chunks? (2 min+ for practical use?)

3. Need to figure out how to set determine `min_length` or `min_new_tokens` based on input text so that the model doesn't miss a few words. Can do something very rudimentary like input_prompt_word_count*0.8 (_or any simple formula that almost gets it right_)

4. The model occasionally just generates screeching noises with no coherent words. Need to determine why.

**Comment by ylacombe** [comment]
2. Ideas on how to generate consistently long vocals based on speaker_id in chunks? (2 min+ for practical use?)

This needs a bit of hacking around the model, i.e by adding some names to indicate to your model to keep voice consistency.
